### PR TITLE
react-packager: Add ES6 import statement support to DependencyGraph.

### PR DESCRIPTION
**Pull request by pilwon**
This PR teaches packager's `DependencyGraph` how to extract dependencies written with ES6 `import` statements.

It fixes the issue where you are not able to write your app with ES6 `import` statements when your custom transformer (replacing the default [JSTransform](https://github.com/facebook/jstransform), for example, [babel](http://babeljs.io/)) already supports the ES6 `import` syntax.

It will also be useful for [JSTransform](https://github.com/facebook/jstransform) later on once it implements `import` feature too.

### CODE REVIEW

**Comment by jgable** [comment]
Does this just work for `import blah from 'blah';`?  Would it work for all the proper import syntaxes?

**Comment by pilwon** [comment]
@jgable Yes, it supports all valid ES6 import syntaxes documented in the [ES6 Language Specification](https://people.mozilla.org/~jorendorff/es6-draft.html#sec-imports).

You can see the result from a test ran against a comprehensive list of valid syntaxes in [this gist](https://gist.github.com/pilwon/ff55634a29bb4456e0dd).

**Comment by amasad** [comment]
We also need to make the change here https://github.com/pilwon/react-native/blob/packager-es6-import-patch/packager/react-packager/src/DependencyResolver/haste/index.js#L28
It'd be great if we can consolidate them.

**Comment by pilwon** [comment]
@amasad Sure I can add that to PR. I have a question:

Why does `DependencyResolver` use `/require\(['"]([\.\/0-9A-Z_$\-]*)['"]\)/gi` while `DependencyGraph` uses `/\brequire\s*\(\s*[\'"]([^"\']+)["\']\s*\)/g`?

* Isn't the case-insensitive option incorrectly applied? For example, there could be a line that calls a custom function defined as `function ReQuire(str) {...}`. That regex will replace the string passed to that function that has nothing to do with `require()`.

* Is it okay `\b` is not present there? Wouldn't it match any function with suffix `require`? (ex: `myrequire(...)`)

**Comment by pilwon** [comment]
@amasad PR #368 addressed above questions. I merged the latest code into this PR branch. Test passes.

**Comment by amasad** [comment]
Thanks, will take a look shortly.

File: packager/react-packager/src/DependencyResolver/haste/__tests__/HasteDependencyResolver-test.js [hunk 1]
```diff
@@ -254,13 +380,139 @@ describe('HasteDependencyResolver', function() {
...
+        'import Default , { Foo , Bar , Baz as Qux } from "changed" ;',
+        'import Default , { Foo , Bar , Baz as Qux , } from "changed" ;',
+        'import Default , { Foo as Bar , Baz as Qux , Norf } from "changed" ;',
+        'import Default , { Foo as Bar , Baz as Qux , Norf , } from "changed" ;',
+        'import Default , { Foo as Bar , Baz , Qux as Norf } from "changed" ;',
+        'import Default , { Foo as Bar , Baz , Qux as Norf , } from "changed" ;',
+        'import Default , { Foo , Bar as Baz , Qux as Norf } from "changed" ;',
+        'import Default , { Foo , Bar as Baz , Qux as Norf , } from "changed" ;',
+        'import Default , { Foo as Bar , Baz as Qux , Norf as Enuf } from "changed" ;',
+        'import Default , { Foo as Bar , Baz as Qux , Norf as Enuf , } from "changed" ;',
```
Comment (amasad) [inline packager/react-packager/src/DependencyResolver/haste/__tests__/HasteDependencyResolver-test.js]: can you add test where the comma is stuck to the default?

Comment (amasad) [inline packager/react-packager/src/DependencyResolver/haste/__tests__/HasteDependencyResolver-test.js]: and a test where there are no spaces between the named imports and the `{` `}`

**Comment by amasad** [comment]
Looks good, let's just add a couple more tests

**Comment by pilwon** [comment]
@amasad I updated test cases so most of them are in a more widely used spacing styles:

```js
import Default, { Foo, Bar as Baz, Qux as Norf } from 'changed';
```

I still kept some different spacing style cases such as `import 'x' ; `, `import {} from 'x';`, `import {Foo as Bar} from 'x';`, and `import Default , { } from 'x';`.

There's an incorrect statement `import'x'; ` that is expected to be ignored by the replace regex.

**Comment by amasad** [comment]
Great, thanks for being very response, @pilwon!

**Pull request closed without merging**